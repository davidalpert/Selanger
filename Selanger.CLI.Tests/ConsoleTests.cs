using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using ApprovalTests;
using NUnit.Framework;
using Selanger.CLI.Tests.Helpers;

namespace Selanger.CLI.Tests
{
    public class ConsoleTests
    {
        [Test]
        public void Get_help()
        {
            var args = new string[] {};

            var output = RunProgramAndCaptureOutput(args);

            Console.WriteLine(output);
        }

        [Test]
        public void ProjectsReport_implicit()
        {
            var solutionFileDirectory = GetSolutionFileDirectory();

            var args = new[]
                {
                    solutionFileDirectory.FullName
                };

            var output = RunProgramAndCaptureOutput(args);

            Console.WriteLine(output);
        }

        [Test]
        public void ProjectsReport_explicit()
        {
            var solutionFileDirectory = GetSolutionFileDirectory();

            var args = new[]
                {
                    "-p",
                    solutionFileDirectory.FullName
                };

            var output = RunProgramAndCaptureOutput(args);

            Console.WriteLine(output);
        }

        [Test]
        public void StatisticsReport()
        {
            var solutionFileDirectory = GetSolutionFileDirectory();

            var args = new[]
                {
                    "-s",
                    solutionFileDirectory.FullName
                };

            var output = RunProgramAndCaptureOutput(args);

            Console.WriteLine(output);
        }

        private static string RunProgramAndCaptureOutput(string[] args)
        {
            string output;
            using (var cons = new RedirectedConsole())
            {
                Program.Main(args);

                output = cons.Output;
            }
            return output;
        }

        private DirectoryInfo GetThisFileDirectory()
        {
            var stacktrace = new StackTrace(true);
            var first_frame = stacktrace.GetFrame(0);
            var path_to_source = first_frame.GetFileName();
            var path_to_source_folder = Path.GetDirectoryName(path_to_source);
            return new DirectoryInfo(path_to_source_folder);
        }

        private static FileInfo GetSelengerExe()
        {
            var directory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
            var path_to_cli = Path.Combine(directory, "Selanger.exe");
            var cli = new FileInfo(path_to_cli);
            Assert.True(cli.Exists, cli.FullName);
            return cli;
        }

        private DirectoryInfo GetSolutionFileDirectory()
        {
            var source_dir = GetThisFileDirectory();
            var path_to_sln = Path.Combine(source_dir.FullName, "..", "Selanger.sln");
            var sln_file = new FileInfo(path_to_sln);
            Assert.True(sln_file.Exists, sln_file.FullName);
            return sln_file.Directory;
        }
    }
}
