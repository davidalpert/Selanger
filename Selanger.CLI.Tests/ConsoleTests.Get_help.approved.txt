Selanger [options] -i solution1.sln solution2.sln ...

Options:
-r|report {reportType} - configures the report type

Report Types (default: projectGraph)
 n|namespaces - list the namespaces in the included projects
 r|references - list the assembly referenced by the included projects
 pg|projectGraph - dump a graph of the solution/project structure

